edsnlp/data/converters.py

```diff
@@ -1080,6 +1082,479 @@ def __call__(self, doc: Doc) -> str:
+            warnings.warn(
+                f"Length mismatch between tokens ({n_tokens}) and ner_tags "
+                f"({len(ner_tags)}); using min length."
+            )
+
+        L = min(n_tokens, len(ner_tags))
+        current_type: Optional[str] = None
+        start_idx: Optional[int] = None
+
+        for i in range(L):
+            raw = self._resolve_label(ner_tags[i])
+            if raw in ("O", "0"):
+                if current_type is not None:
+                    current_type = self._close_entity(doc, entities, current_type, start_idx, i - 1)
+                    start_idx = None
+                continue
+
+            prefix, etype = self._split_tag(raw)
+
+            if prefix in ("B", "S") or current_type is None:
+                current_type, start_idx = self._handle_b_or_s(doc, entities, prefix, etype, current_type, start_idx, i)
+                continue
+
+            if prefix in ("I", "E"):
+                current_type, start_idx = self._handle_i_or_e(doc, entities, prefix, etype, current_type, start_idx, i)
+                continue
+
+            # No explicit prefix: group consecutive same-type tokens
+            current_type, start_idx = self._handle_no_prefix(doc, entities, etype, current_type, start_idx, i)
+
+        if current_type is not None:
+            self._close_entity(doc, entities, current_type, start_idx, L - 1)
+
+        return entities
+
+    def __call__(
+        self,
+        obj: Mapping[str, Any],
+        tokenizer: Optional[Tokenizer] = None,
+    ) -> Doc:
+        tok = tokenizer or self.tokenizer or get_current_tokenizer()
+
+        # Get data from the example
+        words = obj[self.words_column]
+        ner_tags = obj[self.ner_tags_column]
+
+        # Build a spaCy Doc directly from words to avoid tokenizer mismatches
+        vocab = tok.vocab
+        # assume single-space separation between tokens
+        spaces = [True] * (len(words) - 1) + [False]
+        doc = Doc(vocab, words=words, spaces=spaces)
+
+        # Set document ID: prefer configured id column, otherwise use
+        # the file-name sentinel if present.
+        if self.id_column is not None:
+            doc._.note_id = obj.get(self.id_column, obj.get(FILENAME))
+        else:
+            doc._.note_id = obj.get(FILENAME)
+
+        # Extract entities using the token-aligned Doc
+        entities = self._extract_entities(doc, ner_tags)
+
+        # Add entities to doc
+        spans: List[Span] = []
+        for entity in entities:
+            span = doc.char_span(
+                entity["begin"],
+                entity["end"],
+                label=entity["label"],
+                alignment_mode="expand",
+            )
+
+            if span is not None:
+                spans.append(span)
+            else:
+                warnings.warn(
+                    f"Could not align entity '{entity['text']}' "
+                    f"at ({entity['begin']}, {entity['end']})"
+                )
```

Comment:
It's actually simpler to deal with token indices directly, instead of using char_span

```suggestion
        for begin, end, label in entities:
            spans.append(Span(doc, begin, end, label=label))
```